### DISCUSSION TITLE
Fast/recommended path for per-frame buffer uploads?

### DISCUSSION THREAD

**Discussion by mrshannon** [gpuweb/gpuweb#1428] [Q&A]
So I have been digging though the old issues trying to determine what the optimal/recommended path is for per frame buffer uploads (i.e. for MVP/normal matrices and dynamic vertex arrays). I can identify three paths to take. NOTE: I am currently assuming non-unified memory as there does not seem to be a method of determining when staging buffers are not necessary, therefore I am assuming `MAP_WRITE` and `VERTEX/UNIFORM` etc are mutually exclusive.

Create the staging buffers every frame with `createBuffers` and use `mappedAtCreation`, destroy the buffer after the copy to the GPU private buffer is complete. Unless there is some management of allocation behind the scenes (which is not mentioned in the spec) this seems like a bad idea because of the constant allocation/deallocation. As far as I can tell it's not like OpenGL where buffer orphaning is a fast path.

Issue a direct write into the GPU private buffer with `writeBuffer`. This seems very similar to `glBufferSubData` which is controversial as it is a fast path on NVIDIA but is a slow path on Intel. Not sure if that is the case with WebGPU or not as it's actual behavior is not in the spec.

Map (using `mapAsync`) and unmap round robin staging buffers every frame. The difficulty of this is guessing which ones are required for the frame as it's an async operation and thus cannot be performed inside the main render function. It's not impossible, just difficult in some circumstances, though the ability to map a sub range synchronously once the `Promise` is resolved partially negates this. The bigger issue is that mapping and unmapping is traditionally slow. Going back to OpenGL, NVIDIA does not recommend this approach, though Intel seems to prefer it over direct writes like `glBufferSubData`.

In other API's the recommendation from all vendors is to use persistent buffer mapping as it is the fast path for all IHVs. This is neither currently supported nor does there seem to be any plans of supporting this in WebGPU (am I wrong?). Therefore, is there any recommendations on what should be the recommended path on WebGPU, especially since there is no way to determine which IHV it's running on and adjust accordingly?

**Comment by kvark** [comment]
Generally I would recommend `write_buffer`. The performance woes of `glBufferSubData` do not apply to it: when designing WebGPU we are taking extra consideration for performance characteristics to be portable, not just the function. On a high level, what `write_buffer` does is finding staging space internally, filling it, and issuing a `copy_buffer_to_buffer` on the queue. So it's not blocking, it's queued like everything else on the queue, and it's fairly efficient, since the WebGPU implementation has more tools to manage that staging space nicely than the user.

However, under the following conditions you may get better performance by doing a "staging belt" yourself:
  - if your data is *not* already in an array buffer. With mapping, you can serialize it right into the mapped range. With `write_buffer`, you'd need an extra copy on your side, since `write_buffer` expects an array buffer.
  - you are not accessing WebGPU from WASM. Since WASM today can't share memory, it has to do a copy for buffer mapping, but this isn't a problem for `write_buffer`.
  - you are not afraid of additional complexity in the logic. Building a [staging belt](https://github.com/gfx-rs/wgpu-rs/blob/master/src/util/belt.rs) is not rocket science, but there are certainly more places to screw up than with `write_buffer`.

Edit: if you go this path, there is also an extra benefit that you can overwrite sections of your target buffers multiple times in the same submission. With `write_buffer` you can't update data in the middle of a submission (unless doing an extra copy from some place).

> The bigger issue is that mapping and unmapping is traditionally slow

The actual native buffer can be persistently mapped for all you know, so don't worry about this.

> The difficulty of this is guessing which ones are required for the frame as it's an async operation and thus cannot be performed inside the main render function

Not sure I understand that difficulty/guessing. You maintain a pool of staging buffers that are mapped, and once you use any of them, you request it for mapping again, and in the promise you bring it back into the pool. There shouldn't be any guesswork here.

> Create the staging buffers every frame with createBuffers and use mappedAtCreation

This is strictly inferior to `write_buffer`, since the latter knows exactly the lifetime of that staging buffer, while if you are doing that, the implementation has no idea how long this buffer will live, and the allocation and memory reuse aren't going to be as good.

**Reply by Kangz** [reply]
> Not sure I understand that difficulty/guessing. You maintain a pool of staging buffers that are mapped, and once you use any of them, you request it for mapping again, and in the promise you bring it back into the pool. There shouldn't be any guesswork here.

And if you ever run out of space you can instantly create an additional mapped chunk using `mappedAtCreation`. That one chunk will be potentially less efficient the first time it is used, but follow-up mapping will be as efficient as the other chunks in your staging belt.

**Reply by kainino0x** [reply]
It's a bit old but there is a sketch of this idea here: https://github.com/gpuweb/gpuweb/blob/main/design/BufferOperations.md#updating-data-to-an-existing-buffer-like-webgls-buffersubdata

**Comment by Morglod** [comment]
> The actual native buffer can be persistently mapped for all you know, so don't worry about this.
> 
> >

year 2024, macos m1 wgpu-rs implementation, metal backend
mapping/unmapping buffers still slow as hell (as it always was) (16-32ms for 272 bytes of data)

**Reply by greggman** [reply]
Keep some mapped staging buffers around?

pseudo code

```js
const mappedStagingBuffers = [];
function getMappedStagingBuffer() {
  const buffer = mappedStagingBuffers.pop();
  if (buffer) {
    return buffer;
  }
  return device.createBuffer({size, usage: COPY_SRC|MAP_WRITE, mappedAtCreation: true});
}

// When you need a staging buffer
const buffer = getMappedStagingBuffer();  // buffer is already mapped
const data = buffer.getMappedRange();

...put data in...

buffer.unmap();

... use buffer for whatever ...


device.queue.submit([commandBufferThatUsesBuffer]);

// asynchronously remap buffer
buffer.mapAsync(GPUMapMode.WRITE).then(() => mappedStagingBuffers.push(buffer));
```

You should always have a buffer immediately following this pattern. You'll need to decide on how to bucket buffers by size and/or usage

**Reply by Morglod** [reply]
For writing I can use wgpuQueueWriteBuffer. The problem is reading from it. Because its (not possible?) to implement api that may be optimized for each vendor/driver (just read* method, same as write*), we have only map* solution.
So now we should somehow guess how it works under the hood for each vendor/driver/platform, to make it fast.
And we cant leave mapped CPU buffer, it should be unmapped before any operation. I know, its because some vendors don't have persistent mapping, but we also don't have read* method, so now we have only one bad api for reading..

And surprise... its really hard to guess what way is the fastest and why. Tried a lot of solutions, current one is this:

```cpp
// pass some work to encoder

readBuffer = getReadBuffer(); /* ring buffer or createBuffer({
    .mappedAtCreation = false,
    .size = 123,
    .usage = WGPUBufferUsage_CopyDst | WGPUBufferUsage_MapRead
}); */
copyBufferToBuffer(gpuOutput, readBuffer);
queueSubmit();

wgpuBufferMapAsync(WGPUMapMode_Read, readBuffer);
wgpuDevicePoll();
mem = getConstMappedRange(readBuffer);
// do smth with mem
unmap();
```

Tried ring buffer and creating/deleting every frame. When I create/delete every frame, after +-20 frames timings for reading going from 1.2ms to 20ms. Pretty same for ring buffer variant.

Please add read* method and leave implementation for each vendor. Because end-user of this api does not want to do mapping (which also may not be persistent). It makes no sense at all. End user want to read or to write some data. So api should have 4 methods for this: buffer.beginMemoryOp(); buffer.read/write; buffer.endMemoryOp();
Thats all. Then each vendor may have ring buffers/polls/global_buffer or whatever they want and whatever will be the fastest way.

Real code:
<details>

```cpp
    queue.writeBuffer(cam.input_update_buffer, 0, &update_data, sizeof(EcsCameraUpdateData));

    auto encoder = cam.device.createCommandEncoder();
    auto compute_pass = encoder.beginComputePass();

    compute_pass.setPipeline(cam.compute_pipeline);
    compute_pass.setBindGroup(0, cam.bind_group, 0, nullptr);
    compute_pass.dispatchWorkgroups(1, 1, 1);
    compute_pass.end();

    auto current_read_buf = cam.device.createBuffer({
        .mappedAtCreation = false,
        .size = sizeof(EcsCameraData),
        .usage = WGPUBufferUsage_CopyDst | WGPUBufferUsage_MapRead
    });

    encoder.copyBufferToBuffer(cam.output_buffer, 0, current_read_buf, 0, sizeof(EcsCameraData));
    auto commands = encoder.finish();

    queue.submit({ commands });

    commands.release();
    encoder.release();
    compute_pass.release();

    wgpuBufferMapAsync(current_read_buf, mode, offset, size, [](WGPUBufferMapAsyncStatus status, void* user_data) {}, nullptr);
    wgpuDevicePoll(device, true, NULL);

    auto data = (const EcsCameraData*)current_read_buf.getConstMappedRange(0, sizeof(EcsCameraData));
    memcpy(&cam, data, sizeof(EcsCameraData));
    current_read_buf.unmap();
    wgpuBufferDestroy(current_read_buf);

    wgpuQueueRelease(queue);
```

</details>

**Reply by greggman** [reply]
So here's an example doing picking. It draws to a texture, uses `copyTextreToBuffer` to read the texture to buffer the size of the canvas and maps it. I'm getting 0.05ms on M1 Mac and 0.005ms on Win 11 Nvidia 2070 super

https://jsgist.org/?src=e20f28aab7427c68dc4f427c6dc32654

Note that I'm on Chrome, using dawn, and it's probably using `WGPUFutures` which might not be available yet in wgpu? 

> Tried ring buffer and creating/deleting every frame

The pseudo code above only created a new buffer if there wasn't one available. It did not create a new buffer every frame.

**Reply by kainino0x** [reply]
Chrome is not using WGPUFuture yet, but also we have a latency problem in Chrome that won't be immediately fixed by them.

I am surprised by the speed of mapAsync though! I thought just a plain dumb round trip took almost 1ms in chrome - but I guess that was a GL problem.

**Reply by Morglod** [reply]
> It did not create a new buffer every frame

Me too for ring buffer variant.

Looks like its a problem of current wgpu-rs implementation, will try dawn, thank you for this example. My point is that its strange api decision which leads to this kind of problems. API should fit user needs, not implementations. I cant imagine that someone will say "I need to map buffer memory". All you need is: write, write chunks, read, read chunks (chunks - I mean some kind of streaming). (reads and writes may be async)

**Reply by Morglod** [reply]
Yep, after 6 hours of fighting with google's implementation, dawn's mapAsync is 300x times faster. But I still dont understand why I should do mapping, polling and all other stuff instead of just read* call (or 3 calls like begin_read, read, read_end).

@greggman

**Reply by greggman** [reply]
>  mapping, polling and all other stuff instead of just read* cal

requesting a mapping and then waiting for the map to happen is because the buffer might be in use and you need to wait for it to finish being used. *polling* should not happen and I think that's the point of the `WGPUFuture` thing that's being worked on: https://github.com/webgpu-native/webgpu-headers/issues/199  I don't know if that will actually make your code any simpler. I think the point of map is that if you want efficient code, you should ask for the buffer to be mapped, then go do other stuff since the GPU might still be busy with the buffer.

I'm not sure I follow your suggestion of `begin_read`, `read`, `read_end`. I guess you want `begin_read` to halt the CPU and wait for the GPU (which is what your polling is currently doing).

It sounds like maybe you want `mapSync`? (which is being discussed here: https://github.com/gpuweb/gpuweb/issues/2217)

**Reply by ferasboulala** [reply]
~Year 2025, M1 Max, Metal backend, dawn native: Around 500us to map 256 bytes (not that the size of the buffer matters). Note that if I don't map/unmap and instead just ignore, the 500us disappears. This isn't a case of "the gpu is still using it", it's just plain slow.~

Disregard my comments. This is just noise because I am learning. The map/unmap is definitely slow but maybe I am misusing it. It's also possible that it is due to security concerns that the API is built that way.

**Reply by Morglod** [reply]
> possible that it is due to security concerns that the API is built that way.

Reading is possible anyway with mapping or without it. From api perspective you actually dont need it at all. Because what you actually need is reading and writing between client side buffer and gpu buffer (sometimes frequently, every frame, sometimes after operation is done in case of gpgpu). There is absolutely no need in mapping from security or usage perspective and it worked perfectly with old api.
Even in case of async stuff, reading/writing operations could be queued, but implementation defined so driver could optimize it (by driver I mean specific browser or native target)
